### ISSUE TITLE
Include Application Health Check (Revert #3781)

### ISSUE THREAD

**Issue by gnunn1** [argoproj/argo-cd#16870]
Checklist:

* [X] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [X] I've included steps to reproduce the bug.
* [X] I've pasted the output of `argocd version`.

**Describe the bug**

Argo CD no longer includes an Application health check OOTB as of issue #3781, this means whenever a user attempts to use the App-of-App patterns with sync waves it will fail. 

There is a fairly regular cadence of questions about this on the argocd slack channel either were someone asks about why the sync wave isn't working or someone is explaining how Argo CD sync waves work with App-of-App and has to explicitly caveat the missing health check.

**To Reproduce**

Deploy an App-of-App with sync waves , Argo CD will not respect the sync wave since there is no health check in play.

**Expected behavior**

App-of-App should work out of the box with sync waves, the original issue #3781 advocated removing the health check so that the health check would not propagate upwards however IMHO this is the wrong approach for the following reasons:

1. The convention in Argo CD is that the Application health status reflects the health status of the resources it manages. In an App-of-App scenario without the health check this convention is no longer respected.

2. Removing the health check puts on the onus on users to put in the health check to get sync waves to work with App-of-App. IMHO the onus should be on users who want to remove the health check to explicitly define an override for the health check that returns true. 

3. It removes the ability to maintain and improve the Application health check in a central source, while it is referenced in the docs many folks won't bother to update the docs if they have an improvement version of the health check.

As per #2. including the health check does not mean that users who do not want it cannot remove it, they can simply override the health check in their argocd-cm to always return true for Application objects.

If version compatibility is an issue, this change could be introduced in Argo CD 3.0.

**Version**

All versions since 1.8 when this check was removed.

**Comment by mtrin** [comment]
I agree with this - I think it should follow the truth as much as possible and if I'm reading this right, it feels like an important feature was disabled where it could be more of a UI issue

**Comment by ChristianCiach** [comment]
We have been using ArgoCD for two days and already encountered this issue. I am not really comfortable putting a bunch of Lua into my ConfigMap for something that I would expect to work out-of-the-box. It seems wrong for an Application to not reflect the health of the managed resources by default.

**Comment by andrii-korotkov-verkada** [comment]
ArgoCD versions 2.10 and below have reached EOL. Can you upgrade and let us know if the issue is still present, please?

**Comment by andrii-korotkov-verkada** [comment]
This looks like a tradeoff. Why don't you want to put some lua in the configmap though?

**Comment by rouke-broersma** [comment]
Because it is then a custom health check, instead of a health check that follows the argocd ALM process and that can be updated by the argocd developers as insights and usage of argocd changes. In my team we copied the old Argocd Application health check and modified it heavily. There is no way to bring those changes we made back upstream so we can share them with the wide Argocd community. These changes now remain mostly closed source, benefitting only a small subset of users.

**Comment by andrii-korotkov-verkada** [comment]
> and modified it heavily

Can you share the result, please? I need more context, but maybe with a new healthcheck we can bring it back.

**Comment by rouke-broersma** [comment]
> > and modified it heavily
> 
> Can you share the result, please? I need more context, but maybe with a new healthcheck we can bring it back.

A result is here, but that is not the latest version:

https://github.com/argoproj/argo-cd/issues/10550

**Comment by gnunn1** [comment]
> This looks like a tradeoff. Why don't you want to put some lua in the configmap though?

I outlined the reasons in the issue report above. I mean this question could have been asked for the original issue that removed it. "Why don't you want to put some Lua in the configmap to override the default?"

IMHO the default should follow the intuitive path which is for Argo CD to provide health checks for its own CRs. Users expect Applications to be properly checked for health and are surprised when that isn't the case.

Other then that the reasoning I put in the original post feels logical to me.

**Comment by ChristianCiach** [comment]
@gnunn1 While I agree with you, we should keep in mind that you currently cannot use Lua to *remove* a bundled health check. You can only add or override bundled health checks, but you can't disable them.

Since we now know that there are people who don't want health checks for `Application` resources, IMHO the bundled health check should only be restored after this PR has been merged:

* https://github.com/argoproj/argo-cd/pull/20462

**Comment by andrii-korotkov-verkada** [comment]
I'll bring this up at the next contributors meeting.

**Comment by andrii-korotkov-verkada** [comment]
Meanwhile, can you check if the result mentioned in the issue is what we'd want and if ore up-to-date is needed, paste it, please?

**Comment by drewburr** [comment]
> @gnunn1 While I agree with you, we should keep in mind that you currently cannot use Lua to _remove_ a bundled health check. You can only add or override bundled health checks, but you can't disable them.
> 
> Since we now know that there are people who don't want health checks for `Application` resources, IMHO the bundled health check should only be restored after this PR has been merged:
> 
> * [feat(app): add ignore-healthcheck annotation #20462](https://github.com/argoproj/argo-cd/pull/20462)

Wanted to resurface this conversation since #20462 has been merged. For continuity, I'm using the below health check from #10550 and see expected behavior when using the app of apps pattern; being that all resources in an application must report healthy before the app reports healthy and can progress the sync wave.

```yaml
resource.customizations.useOpenLibs.argoproj.io_Application: true
resource.customizations.health.argoproj.io_Application: |
  hs = {}
  hs.status = "Progressing"
  hs.message = ""
  if obj.status ~= nil then
    local status = obj.status
    if status.conditions ~= nil then
      for i, condition in ipairs(status.conditions) do
        if condition.type ~= nil and string.match(condition.type, '.*Error$') then
          hs.status = "Degraded"
          hs.message = condition.message
          return hs
        end
      end
    end
    if status.health ~= nil then
      local health = status.health
      hs.status = health.status
      if health.message ~= nil then
        hs.message = health.message
      end
      local syncStatus = (status.sync and status.sync.status or nil)
      if hs.status == "Healthy" and syncStatus ~= "Synced" then
        hs.status = "Progressing"
      end
    end
  end
  return hs
```

**Comment by andrii-korotkov-verkada** [comment]
Discussed at the contributors meeting. Application health check has non-trivial implications to the eventual consistency model. Overall, people are not against this idea, but someone has to come up with a health check that works in all circumstances for existing Argo CD architecture.

**Comment by BenjaminNeale-Heritage** [comment]
Does the existence of the "ignore-healthcheck" annotation [1] make it easier to re-add the `Application` health check?  If there were situations where the `Application` health check was causing issues desired then the "ignore-healthcheck" annotation could be added to `Application` resource(s) to achieve the current state of no `Application` health checks.

Argo CD 3.0 might be a good opportunity to add the `Application` health check back into Argo CD.

> Discussed at the contributors meeting. Application health check has non-trivial implications to the eventual consistency model. Overall, people are not against this idea, but someone has to come up with a health check that works in all circumstances for existing Argo CD architecture.

This is probably the biggest challenge to overcome.  From what you've said, it sounds like there are some curly situations that need to be considered.  However, it may not be reasonable for a solution to first identify 'all circumstances' before being implemented.  What would represent 'close enough'?


[1] https://github.com/argoproj/argo-cd/pull/20462

**Comment by jcpunk** [comment]
Due to the complexity implications, could the example health check be added in, but only run against `Application` with a specific annotation/label?

I'm showing https://github.com/argoproj/argo-cd/issues/10550 has a different lua than is linked here. As different folks develop different scripts, if we have a "bundled rule" that at least provides a target to file pull requests against.

Different workflows probably have weird edges, but it we make the health checks opt-in that would theoretically make it easier to start identifying what edge cases are out in the wild.
